### DISCUSSION TITLE
Multiple meson.options files

### DISCUSSION THREAD

**Discussion by gerioldman** [mesonbuild/meson#12240] [Ideas]
Hi,
I am not sure this is an existing feature, or not, so I put this in the idea category.

I am trying to move an automotive embedded project from a cobbled together Make+Java build system to Meson. Part of the build system is the usage of variants, in Meson terms, options. Most projects at the company will compile different configurations of the same source files, the only change is the selected options. This is done to support more than one hardware.

Now, the way I would think to go about this in Meson is to use multiple meson.options files, but as far as I know, there is no way to select one from the command line, or load it from a subdir.(It's a monorepo, so I am not using the subproject feature)

If this really does not exist, I think a command line option when configuring the folder would be nice, for example:

meson configure builddir --option-file=Config27.options

**Comment by eli-schwartz** [comment]
The options file describes what options are available to select, rather that what options the current configuration has selected. Probably you want to do one of the following instead:
- pass a number of `-Doption1=value1` command line parameters to select the options that the current overall configuration uses
- use a machine file to specify the values for each option: https://mesonbuild.com/Machine-files.html#project-specific-options

My suspicion is you want to use the latter. You could have e.g.
- profiles/config1.ini
- profiles/config2.ini
- ...
- profiles/config27.ini

And select a specific profile with `meson setup builddir --native-file profiles/config27.ini`

Note that machine files can set many things, not just the value of a project option. You can define which compiler to use by full path, select the ld implementation to use (perhaps you prefer lld or mold instead of bfd that is often the compiler's own default), toggle built-in options such as debug/optimization/LTO, set additional directories to have pkg-config search for dependencies in, and more.

**Reply by gerioldman** [reply]
I had the impression that the machine files were more for configuration of the compiler and the linker, etc. In this case, the configs have the same machine file configuration,  compiler, flags, and all. The only difference is in software features, that's why I was asking about CLI selectable options file support. I wanted to separate the compiler options, and the software feature options. As far as I know, you can already set value of the options in the meson.options file, I was not aware you can set the value of options in the machine file.

I would still prefer to separate the two, I feel it makes sense to separate configuration, for example, compiler configuration for a certain MCU, and the configuration of the software features, like what control algorithm the software uses.

Is there any problem that could be caused by user selectable options file? I can only think of conflict between the machine and the options file, like setting different value for the same option, but that already exists.

**Reply by gerioldman** [reply]
I forgot to reply to this part:
-Doption1=value1

In my case, we are talking about ~30-100 options, so setting options by command line arguments is kind of ruled out.

**Reply by eli-schwartz** [reply]
The options file only defines the default if not manually specified. Allowing multiple options files could cause issues because meson won't function without one in the expected location -- you cannot even run `meson configure` to get the help text for various options, and running meson but forgetting to specify the options file would unpredictably error out with messages saying that `meson.build` tried to "access an unknown option". I would be hesitant to generally allow it. Plus, you'd have to repeat the type information and possibly description for each option across every copy of the file.

I think that machine files are, at the generic level, a better solution for this.

> I wanted to separate the compiler options, and the software feature options.

Does it help if I note that you can use multiple machine files and they will be layered and their options merged? You can have a machine file defining the compiler settings, and a different set of machine files that define a set of options composing a hardware configuration. Run meson with both:
- the compiler machine file
- one of a collection of hardware configuration machine files

and the resulting build will use all the settings from both files.

**Reply by gerioldman** [reply]
Yes, I was reading the Meson website, and I've seen that you can specify more than one machine file after writing the response. I was not aware of this. I think this will be the best way to handle this, thanks for the help!
